LogMonitor/src/LogMonitor/FileMonitor/Utilities.cpp

```diff
@@ -81,82 +82,131 @@ HANDLE GetLogDirHandle(std::wstring logDirectory, HANDLE stopEvent) {
-                    //Wait failed, return the failure.
-                    //
-                    status = GetLastError();
-
-                    CancelWaitableTimer(timerEvent);
-                    CloseHandle(timerEvent);
-
-                    return INVALID_HANDLE_VALUE;
-                }
-            }
-
-            logDirHandle = CreateFileW (logDirectory.c_str(),
-                                          FILE_LIST_DIRECTORY,
-                                          FILE_SHARE_READ | FILE_SHARE_WRITE,
-                                          nullptr,
-                                          OPEN_EXISTING,
-                                          FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
-                                          nullptr);
-
-            if (logDirHandle == INVALID_HANDLE_VALUE)
-            {
-                status = GetLastError();
-                ++retryCount;
-            }
-            else
-            {
-                status = ERROR_SUCCESS;
-                break;
-            }
+            status = GetLastError();
+            logWriter.TraceError(
+                Utility::FormatString(
+                    L"Failed to set timer object to monitor log file changes in directory %s. Error: %lu",
+                    logDirectory.c_str(),
+                    status)
+                    .c_str());
+            break;
         }
 
-        CancelWaitableTimer(timerEvent);
-        CloseHandle(timerEvent);
+        DWORD wait = WaitForMultipleObjects(eventsCount, dirOpenEvents, FALSE, INFINITE);
+        switch (wait)
+        {
+        case WAIT_OBJECT_0:
+        {
+            //
+            // The process is exiting. Stop the timer and return.
+            //
+            CancelWaitableTimer(timerEvent);
+            CloseHandle(timerEvent);
+            return INVALID_HANDLE_VALUE;
+        }
+
+        case WAIT_OBJECT_0 + 1:
+        {
+            //
+            // Timer event. Retry opening directory handle.
+            //
+            break;
+        }
+
+        default:
+        {
+            //
+            // Wait failed, return the failure.
+            //
+            status = GetLastError();
+
+            CancelWaitableTimer(timerEvent);
+            CloseHandle(timerEvent);
+
+            return INVALID_HANDLE_VALUE;
+        }
+        }
+
+        logDirHandle = CreateFileW(logDirectory.c_str(),
+                                   FILE_LIST_DIRECTORY,
+                                   FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                   nullptr,
+                                   OPEN_EXISTING,
+                                   FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
+                                   nullptr);
+
+        if (logDirHandle == INVALID_HANDLE_VALUE)
+        {
+            status = GetLastError();
+            elapsedTime += WAIT_INTERVAL;
+        }
+        else
+        {
+            status = ERROR_SUCCESS;
```

Comment:
Once the directory is detected after the several interval timed checks, would it be helpful to print the success message on STDOUT? 

For example: If the directory is created within the wait time and no changes are detected in .log files within the directory, the blank screen might confuse the user into thinking that the tool has not detected the new directory yet.